src/main/java/com/google/sps/servlets/ReviewsServlet.java

```diff
@@ -0,0 +1,136 @@
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import static com.google.sps.Constants.*;
+import com.google.sps.Review;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Servlet that handles customer reviews.
+ */
+@WebServlet("/reviews")
+public class ReviewsServlet extends HttpServlet {
+  Gson gson = new Gson();
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  /** Get review from review form and store it in Datastore under its Business entity. */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) {
+    try {
+      // Get review input from review form.
+      String userID = getParameter(request, "name", "anonymous"); // Accept input name for now, should be user Datastore key, to be implemented later
+      String comment = getParameter(request, "review", "No comment provided.");
+      int rating = Integer.parseInt(request.getParameter("rating"));
+
+      // Get current timestamp for time when review is posted.
+      LocalDateTime dateTimeObj = LocalDateTime.now();
+      DateTimeFormatter format = DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm:ss");
+      String dateTime = dateTimeObj.format(format);
+
+      // Get existing reviews key list.
+      Key businessKey = KeyFactory.createKey("Business", Long.parseLong(request.getParameter(BUSINESS_KEY))); // KIV url parameter
+      Entity businessEntity = datastore.get(businessKey);
+      Key[] reviewsKeyArr = gson.fromJson((String) businessEntity.getProperty(BUSINESS_REVIEWS), Key[].class);
+      List<Key> reviewsKeyList = new ArrayList<Key>(Arrays.asList(reviewsKeyArr));
+
+      // Create review entity and get its key.
+      Entity reviewEntity = new Entity("Review");
+      reviewEntity.setProperty("userID", userID);
+      reviewEntity.setProperty("comment", comment);
+      reviewEntity.setProperty("rating", rating);
+      reviewEntity.setProperty("dateTime", dateTime);
+      datastore.put(reviewEntity);
+      Key reviewKey = reviewEntity.getKey();
+
+      // Append key to its Business entity.
+      reviewsKeyList.add(reviewKey);
+      businessEntity.setProperty(BUSINESS_REVIEWS, gson.toJson(reviewsKeyList));
+
+      // Redirect back to the product page.
+      response.sendRedirect("/product.html?businessID="+businessKey);
```

Comment:
fun fact! you can use template literals like this:

`response.sendRedirect(`/product.html?businessID=${businessKey}`);`